subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -332,6 +333,10 @@ static struct lwm2m_engine_obj_inst *firmware_create(uint16_t obj_inst_id)
 			  &update_state, sizeof(update_state));
 	INIT_OBJ_RES_DATA(FIRMWARE_UPDATE_RESULT_ID, res, i, res_inst, j,
 			  &update_result, sizeof(update_result));
+	INIT_OBJ_RES_OPTDATA(FIRMWARE_PACKAGE_NAME_ID, res, i, res_inst, j);
+	INIT_OBJ_RES_OPTDATA(FIRMWARE_PACKAGE_VERSION_ID, res, i, res_inst, j);
+	INIT_OBJ_RES_DATA(FIRMWARE_UPDATE_PROTO_SUPPORT_ID, res, i, res_inst, j,
```

Comment:
Shouldn't this also be optional?

Reply:
I had a bit trouble understanding these macros for object/resource creation. 
AFAIK **lwm2m_engine_obj_field** array is used for defining field attributes (read, write, exe and optional/mandatory). 
Macro containig OPT suffix indicates whether resource data will be set during macro call or later point in time with lwm2m_engine_set_res_data? 

I would really appreciate having these macros documented, as I it is completely unclear what does the each one do. This would also simplify custom object creation a lot.

Reply:
Yes, OPT suffix indicate that the resource is optional, and has to be set manually with `lwm2m_engine_set_res_data()` by the application in order to enable it. Otherwise, the resource would always be included with some default value, even if the application is not intended to use it.

Reply:
> Yes, OPT suffix indicate that the resource is optional` 

To clarify, does the optional here facilitate Optional characteristic from lwm2m specs? If that's the case why is the Package resource initiated with INIT_OBJ_RES_OPT? 

My train of thought was following. LwM2M Mandatory/Optional characteristic is set through lwm2m_engine_obj_field array.
These helper macros with suffix OPT are not directly connected to resource being mandatory or optional per spec. 
The only difference between INIT_OBJ_RES_OPTDATA and INIT_OBJ_RES_DATA is in the pointer and data length assignment. So here, OPT indicates whether resource data is set when create callback is called or at later point in time by client code.

In case of FIRMWARE_UPDATE_PROTO_SUPPORT_ID, I did not initialise it with OPT suffix because I am setting data there just like FIRMWARE_UPDATE_RESULT_ID using static variable.

Please correct me if I'm wrong.

Reply:
> If that's the case why is the Package resource initiated with INIT_OBJ_RES_OPT?

I won't deny there are some inconsistencies across different object, not only the Firmware Update, nobody bothered to "fix" them as it's really a minimal impact. I'm just not convinced with the argument, we should introduce new ones, because there are other instances already.

> LwM2M Mandatory/Optional characteristic is set through lwm2m_engine_obj_field array.

That's correct, however, I'd consider `OBJ_FIELD_DATA` macro as a way to define an object model. The mandatory/optional field in this case merely instructs the engine, whether it's ok if we fail to read/write a resource (because it's optional and might not be available) or if an error should rather be reported in such case.

> These helper macros with suffix OPT are not directly connected to the resource being mandatory or optional per spec.
The only difference between INIT_OBJ_RES_OPTDATA and INIT_OBJ_RES_DATA is in the pointer and data length assignment.

Well, we should ask the author probably but what would the "OPT" refer to in this case (I agree the documentation on the internal headers is very poor).

Anyway, from my point of view, the presence of the data pointer at the initialization is the core difference between a mandatory/optional resource. For mandatory resources, the buffer has to be present as not being able to read/write a mandatory resource is an error condition. For optional, you don't provide the buffer by default, so it's simply ignored by the engine. If the application decides to use the optional resource, it should "enable" it by providing data buffer, but it shouldn't be done by default IMO.

I see, however (looking at IPSO objects which contain mostly of optional resources, that it's been a common pattern to initialize optional resources by default... So if you really want to keep the default initialization i won't object anymore, but I really think this should be sorted out in the future (some different way to enable/disable optional resources perhaps?).

Reply:
Let's merge it like this.

I am definitely in favor of documenting internal macros (+ refactoring if necessary) and making their usage consistent across all object implementations. Unfortunately I don't have time to do this now. I checked blame of the lwm2m_object.h and it's mostly old code (written in 1.x zephyr) written by @mike-scott.